pkg/compactor/executor.go

```diff
@@ -283,7 +293,83 @@ func (e *schedulerExecutor) updateJobStatus(ctx context.Context, key *compactors
+		level.Error(e.logger).Log("msg", "invalid compaction plan, abandoning job", "tenant", spec.Tenant)
+		return compactorschedulerpb.ABANDON, errCompactionJobHasNoBlocks
+	}
+
+	userID := spec.Tenant
+	userLogger := util_log.WithUserID(userID, e.logger)
+
+	reg := prometheus.NewRegistry()
+	defer c.syncerMetrics.gatherThanosSyncerMetrics(reg, userLogger)
+
+	userBucket := bucket.NewUserBucketClient(userID, c.bucketClient, c.cfgProvider)
+
+	// Omit ShardAwareDeduplicateFilter. Duplicates are cleaned up during planning via GarbageCollect().
+	// Keep LabelRemoverFilter to remove deprecated labels if applicable.
+	fetcherFilters := []block.MetadataFilter{
+		NewLabelRemoverFilter(compactionIgnoredLabels),
+	}
+
+	var maxLookback = c.cfgProvider.CompactorMaxLookback(userID)
+	if c.cfgProvider.CompactorBlockUploadEnabled(userID) {
+		maxLookback = 0
+	}
+
+	syncDir := c.metaSyncDirForUser(userID)
+	fetcher, err := block.NewMetaFetcher(userLogger, c.compactorCfg.MetaSyncConcurrency, userBucket, syncDir, reg, fetcherFilters, maxLookback)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create meta fetcher")
+	}
+
+	syncer, err := newMetaSyncer(userLogger, reg, userBucket, fetcher, nil, c.blocksMarkedForDeletion)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create syncer")
+	}
+
+	blockIDs := make([]ulid.ULID, 0, len(spec.Job.BlockIds))
+	for _, id := range spec.Job.BlockIds {
+		blockID, err := ulid.Parse(string(id))
+		if err != nil {
+			level.Error(userLogger).Log("msg", "invalid block ID, abandoning job", "tenant", userID, "err", err)
+			return compactorschedulerpb.ABANDON, errors.Wrapf(err, "failed to parse block ID")
+		}
+		blockIDs = append(blockIDs, blockID)
+	}
+
+	if err := syncer.SyncRequestedMetas(ctx, blockIDs); err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to sync metas")
+	}
+
+	jobMetas, err := getJobMetasFromSyncer(syncer, blockIDs)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, err
+	}
+
+	var splitNumShards uint32
+	if spec.Job.Split {
+		splitNumShards = uint32(c.cfgProvider.CompactorSplitAndMergeShards(userID))
+	}
+
+	job, err := buildCompactionJobFromMetas(userID, key.Id, jobMetas, spec, splitNumShards)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, err
+	}
+
+	compactor, err := c.newBucketCompactor(ctx, userID, userLogger, userBucket, syncer, reg)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create bucket compactor")
+	}
+
+	level.Info(userLogger).Log("msg", "executing compaction job from scheduler", "tenant", userID, "blocks", len(blockIDs), "split", spec.Job.Split)
+	_, compactedBlockIDs, err := compactor.runCompactionJob(ctx, job)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, err
+	}
+
+	level.Info(userLogger).Log("msg", "compaction job completed", "tenant", userID, "compacted_blocks", len(compactedBlockIDs))
```

Comment:
In the context of where this gets called in `Compact` there is more to do after with metrics and handling error cases:
https://github.com/grafana/mimir/blob/47d83cc3d82166a10263b937dbaec0f34cd70682/pkg/compactor/bucket_compactor.go#L987

Something that also feels missing is the lifecycle of the files that are downloaded during compaction. I see them being cleaned up as best effort if we encounter an error, but if that encounters failures or we crash halfway I don't see where we're cleaning them up again. `Compact` interestingly doesn't clean up on errors because it thinks it will use the files again: https://github.com/grafana/mimir/blob/47d83cc3d82166a10263b937dbaec0f34cd70682/pkg/compactor/bucket_compactor.go#L938 but in our flow we might never get the same job again after a failure.